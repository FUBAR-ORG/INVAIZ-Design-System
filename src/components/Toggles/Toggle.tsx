import styled from '@emotion/styled';

export interface ToggleProps {
  checked?: boolean;
  onChange?: (checked: boolean) => void;
}

function Toggle({ checked, onChange }: ToggleProps) {
  return (
    <Label>
      <Input type='checkbox' checked={checked} onChange={(e) => onChange?.(e.target.checked)} />
      <Span />
    </Label>
  );
}

const Label = styled.label`
  position: relative;
  display: inline-block;
  width: 48px;
  height: 24px;
  input:checked + span {
    background: ${({ theme }) => theme.color.point1};
  }
  input:checked + span:before {
    transform: translateX(24px);
  }
`;

const Input = styled.input`
  width: 0;
  height: 0;
  opacity: 0;
`;

const Span = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: ${({ theme }) => theme.color.light};
  border-radius: 12px;
  transition: background 0.15s;
  &:before {
    position: absolute;
    content: '';
    width: 20px;
    height: 20px;
    top: 2px;
    left: 2px;
    background: ${({ theme }) => theme.darkColor.basic};
    border-radius: 100%;
    transition: transform 0.15s;
  }
`;

export default Toggle;
