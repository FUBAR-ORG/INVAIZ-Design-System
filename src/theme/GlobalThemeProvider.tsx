import { PropsWithChildren } from 'react';
import { ThemeProvider } from '@emotion/react';
import font, { fontSize, fontWeight, lineHeight } from '@theme/font';
import color from '@theme/color';

export default function GlobalThemeProvider({ children }: PropsWithChildren<unknown>) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

const theme = { font, fontSize, lineHeight, fontWeight, color };
