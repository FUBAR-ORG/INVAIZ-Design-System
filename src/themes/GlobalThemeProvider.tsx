import type { PropsWithChildren } from 'react';
import type { EmotionTheme } from '@themes/emotion-theme';
import type { ColorTheme } from '@themes/colors/color';
// types

import { ThemeProvider } from '@emotion/react';
// React modules

import font, { fontSize, fontWeight, lineHeight, textAlign, fontColor } from '@themes/font';
import { colorMap, normal, light } from '@themes/colors/color';
import style from '@themes/style';
import GlobalStyle from '@themes/GlobalStyle';

interface GlobalThemeProviderProps {
  colorTheme?: ColorTheme;
}

export default function GlobalThemeProvider({
  children,
  colorTheme = 'normal',
}: PropsWithChildren<GlobalThemeProviderProps>) {
  const theme: EmotionTheme = {
    normal,
    light,

    color: colorMap[colorTheme],

    font,
    fontSize,
    lineHeight,
    fontWeight,
    textAlign,
    fontColor,

    style,
  };
  return (
    <ThemeProvider theme={theme}>
      {children}
      <GlobalStyle />
    </ThemeProvider>
  );
}
