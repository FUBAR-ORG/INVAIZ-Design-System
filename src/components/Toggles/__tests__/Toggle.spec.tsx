import Toggle from '@components/Toggles/Toggle';
import { render } from '@tests/test-utils';
import { create } from 'react-test-renderer';
import GlobalThemeProvider from '@themes/GlobalThemeProvider';

describe('Toggle', () => {
  it('토글의 초기값은 false로 설정된다.', () => {
    const { getByRole } = render(<Toggle />);
    expect(getByRole('checkbox')).not.toBeChecked();
  });

  it('토글의 상태를 제어할 수 있다.', () => {
    const { getAllByRole } = render(
      <>
        <Toggle checked={false} />
        <Toggle checked />
      </>
    );
    const [first, second] = getAllByRole('checkbox');
    expect(first).not.toBeChecked();
    expect(second).toBeChecked();
  });

  it('토글을 클릭할 수 있다.', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<Toggle onChange={onChange} />);
    const checkbox = getByRole('checkbox');

    checkbox.click();
    expect(onChange).toHaveBeenCalledWith(true);
    expect(checkbox).toBeChecked();

    checkbox.click();
    expect(onChange).toHaveBeenCalledWith(false);
    expect(checkbox).not.toBeChecked();
  });
});

test('Toggle Snapshot', () => {
  const checked = create(
    <GlobalThemeProvider>
      <Toggle checked />
    </GlobalThemeProvider>
  ).toJSON();
  expect(checked).toMatchSnapshot();

  const unChecked = create(
    <GlobalThemeProvider>
      <Toggle />
    </GlobalThemeProvider>
  ).toJSON();
  expect(unChecked).toMatchSnapshot();
});
