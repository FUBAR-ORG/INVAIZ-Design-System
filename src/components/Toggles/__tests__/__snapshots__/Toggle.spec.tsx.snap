// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Snapshot 1`] = `
<label
  className="css-i1ha2l"
>
  <input
    checked={true}
    className="css-1lckzqk"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="css-cexpd9"
  />
</label>
`;

exports[`Toggle Snapshot 2`] = `
<label
  className="css-i1ha2l"
>
  <input
    className="css-1lckzqk"
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="css-cexpd9"
  />
</label>
`;
